File: fridainjector/src/main/java/com/igio90/fridainjector/FridaInjector.java
package com.igio90.fridainjector;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Build;

import com.topjohnwu.superuser.Shell;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class FridaInjector {
    private final Context mContext;
    private final File mInjector;

    // 静态块初始化 libsu 的配置（可选）
    static {
        Shell.setDefaultBuilder(Shell.Builder.create()
            .setFlags(Shell.FLAG_REDIRECT_STDERR)
            .setTimeout(10));
    }

    private FridaInjector(FridaInjector.Builder builder) {
        mContext = builder.mContext;
        mInjector = builder.getInjector();
    }

    private boolean isProcessRunning(String packageName) {
        // 使用 pidof 检查进程是否存在，这是最快的方式
        List<String> out = Shell.cmd("pidof " + packageName).exec().getOut();
        return !out.isEmpty();
    }

    public void inject(FridaAgent fridaAgent, String packageName, boolean spawn) {
        if (mInjector == null) {
            throw new RuntimeException("did you forget to call init()?");
        }

        if (!isProcessRunning(packageName)) {
            spawn = true;
        }

        StringBuilder agent = new StringBuilder(fridaAgent.getWrappedAgent());

        if (!fridaAgent.getInterfaces().isEmpty()) {
            try {
                ApplicationInfo ownAi = fridaAgent.getPackageManager().getApplicationInfo(
                        fridaAgent.getPackageName(), 0);
                String ownApk = ownAi.publicSourceDir;
                ApplicationInfo targetAi = fridaAgent.getPackageManager().getApplicationInfo(packageName, 0);
                
                File targetApkFile = new File(targetAi.publicSourceDir);
                String targetPath = targetApkFile.getParent();

                // libsu 处理挂载和拷贝更加优雅
                Shell.Job job = Shell.cmd();
                if (targetPath.startsWith("/system/")) {
                    job.add("mount -o remount,rw /system");
                }
                
                job.add("cp " + ownApk + " " + targetPath + "/xd.apk");
                job.add("chmod 644 " + targetPath + "/xd.apk");
                
                if (targetPath.startsWith("/system/")) {
                    job.add("chown root:root " + targetPath + "/xd.apk");
                    job.add("mount -o remount,ro /system");
                } else {
                    job.add("chown system:system " + targetPath + "/xd.apk");
                }
                job.exec();

                agent.append(FridaAgent.sRegisterClassLoaderAgent);

                for (LinkedHashMap.Entry<String, Class<? extends FridaInterface>> entry :
                        fridaAgent.getInterfaces().entrySet()) {
                    agent.append("Java['")
                            .append(entry.getKey())
                            .append("'] = function() {")
                            .append("var defaultClassLoader = Java.classFactory.loader;")
                            .append("Java.classFactory.loader = Java.classFactory['xd_loader'];")
                            .append("var clazz = Java.use('")
                            .append(entry.getValue().getName())
                            .append("').$new();")
                            .append("var args = [];")
                            .append("for (var i=0;i<arguments.length;i++) {")
                            .append("args[i] = arguments[i]")
                            .append("}")
                            .append("clazz.call(Java.array('java.lang.Object', args));")
                            .append("Java.classFactory.loader = defaultClassLoader;")
                            .append("};")
                            .append("\n");
                }
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }

        File fridaAgentFile = new File(fridaAgent.getFilesDir(), "wrapped_agent.js");
        Utils.writeToFile(fridaAgentFile, agent.toString());
        Shell.cmd("chmod 777 " + fridaAgentFile.getPath()).exec();

        if (spawn) {
            Intent launchIntent = mContext.getPackageManager().getLaunchIntentForPackage(packageName);
            // 杀进程
            Shell.cmd("am force-stop " + packageName).exec();
            
            new Thread(() -> {
                long start = System.currentTimeMillis();
                while (!isProcessRunning(packageName)) {
                    try {
                        Thread.sleep(250);
                        if (System.currentTimeMillis() - start > TimeUnit.SECONDS.toMillis(5)) {
                            throw new RuntimeException("wait timeout for process spawn");
                        }
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
                executeInjectCommand(packageName, fridaAgentFile.getPath());
            }).start();

            if (launchIntent != null) {
                mContext.startActivity(launchIntent);
            }
        } else {
            executeInjectCommand(packageName, fridaAgentFile.getPath());
        }
    }

    private void executeInjectCommand(String packageName, String agentPath) {
        Shell.cmd(mInjector.getPath() + " -n " + packageName +
                " -s " + agentPath + " --runtime=v8 -e").exec();
    }

    public static class Builder {
        private final Context mContext;
        private String mArmBinaryPath;
        private String mArm64BinaryPath;
        private String mX86BinaryPath;
        private String mX86_64BinaryPath;
        private File mInjector;

        public Builder(Context context) {
            // libsu 会在第一次调用 Shell.getShell() 时弹出授权申请
            if (!Shell.getShell().isRoot()) {
                throw new RuntimeException("failed to obtain root permissions or device not rooted");
            }
            mContext = context;
        }

        public Builder withArmInjector(String armInjectorBinaryAssetName) {
            mArmBinaryPath = armInjectorBinaryAssetName;
            return this;
        }

        public Builder withArm64Injector(String arm64InjectorBinaryAssetName) {
            mArm64BinaryPath = arm64InjectorBinaryAssetName;
            return this;
        }

        public Builder withX86Injector(String x86InjectorBinaryAssetName) {
            mX86BinaryPath = x86InjectorBinaryAssetName;
            return this;
        }

        public Builder withX86_64Injector(String x86_64InjectorBinaryAssetName) {
            mX86_64BinaryPath = x86_64InjectorBinaryAssetName;
            return this;
        }

        public FridaInjector build() throws IOException {
            if (mArmBinaryPath == null && mArm64BinaryPath == null &&
                    mX86BinaryPath == null && mX86_64BinaryPath == null) {
                throw new RuntimeException("injector asset file name not provided");
            }

            String arch = getArch();
            String injectorName = null;
            switch (arch) {
                case "arm": injectorName = mArmBinaryPath; break;
                case "arm64": injectorName = mArm64BinaryPath; break;
                case "x86": injectorName = mX86BinaryPath; break;
                case "x86_64": injectorName = mX86_64BinaryPath; break;
            }

            if (injectorName == null) {
                throw new RuntimeException("injector binary not provided for arch: " + arch);
            }

            mInjector = extractInjectorIfNeeded(mContext, injectorName);
            return new FridaInjector(this);
        }

        private File getInjector() {
            return mInjector;
        }
    }

    private static File extractInjectorIfNeeded(Context context, String name) throws IOException {
        File injectorPath = new File(context.getFilesDir(), "injector");
        File injector = new File(injectorPath, name);

        if (!injectorPath.exists()) {
            injectorPath.mkdir();
        } else {
            File[] files = injectorPath.listFiles();
            if (files != null && files.length > 0) {
                if (files[0].getName().equals(name)) {
                    return injector;
                }
                files[0].delete();
            }
        }

        Utils.extractAsset(context, name, injector);
        Shell.cmd("chmod 777 " + injector.getPath()).exec();
        return injector;
    }

    private static String getArch() {
        for (String androidArch : Build.SUPPORTED_ABIS) {
            switch (androidArch) {
                case "arm64-v8a": return "arm64";
                case "armeabi-v7a": return "arm";
                case "x86_64": return "x86_64";
                case "x86": return "x86";
            }
        }
        throw new RuntimeException("Unable to determine arch");
    }
}

File: fridainjector/src/main/java/me/voltual/fridainjector/FridaInjector.kt
package me.voltual.fridainjector

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import com.topjohnwu.superuser.Shell
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread
import android.os.Handler
import android.os.Looper

class FridaInjector private constructor(builder: Builder) {
    private val context: Context = builder.context
    private val injector: File = builder.injector ?: throw RuntimeException("Injector not found")

    var loggingCallback: ((String) -> Unit)? = null

    companion object {
        private fun extractInjectorIfNeeded(context: Context, name: String): File {
            val injectorPath = File(context.filesDir, "injector")
            val injector = File(injectorPath, name)
            if (!injectorPath.exists()) injectorPath.mkdir()
            
            Utils.extractAsset(context, name, injector)
            Shell.cmd("chmod 777 ${injector.path}").exec()
            return injector
        }

        private val arch: String
            get() {
                for (androidArch in Build.SUPPORTED_ABIS) {
                    return when (androidArch) {
                        "arm64-v8a" -> "arm64"
                        "armeabi-v7a" -> "arm"
                        "x86_64" -> "x86_64"
                        "x86" -> "x86"
                        else -> continue
                    }
                }
                throw RuntimeException("Unable to determine arch")
            }
    }

    private fun getPid(packageName: String): String? {
        val result = Shell.cmd("pidof $packageName").exec()
        return result.out.firstOrNull()?.trim()
    }

    fun inject(fridaAgent: FridaAgent, packageName: String, spawn: Boolean = false) {
        val agentScript = StringBuilder(fridaAgent.wrappedAgent)
        // TODO: 处理 interfaces (保持不变，此处省略)
        
        val tempAgentPath = "/data/local/tmp/wrapped_agent.js"
        val localFile = File(context.filesDir, "temp_agent.js")
        Utils.writeToFile(localFile, agentScript.toString())
        
        log("[System] 准备环境...")
        Shell.cmd("cp ${localFile.path} $tempAgentPath").exec()
        Shell.cmd("chmod 777 $tempAgentPath").exec()
        Shell.cmd("chmod 777 ${injector.path}").exec()

        log("[System] 放宽 SELinux 策略...")
        Shell.cmd("setenforce 0").exec()

        if (spawn || getPid(packageName) == null) {
            val launchIntent = context.packageManager.getLaunchIntentForPackage(packageName)
            log("[System] 正在重启应用: $packageName")
            Shell.cmd("am force-stop $packageName").exec()

            thread {
                val start = System.currentTimeMillis()
                var pid: String? = null
                while (pid == null) {
                    try {
                        Thread.sleep(500)
                        pid = getPid(packageName)
                        if (System.currentTimeMillis() - start > 10000) {
                            log("[Error] 等待进程启动超时")
                            return@thread
                        }
                    } catch (e: Exception) {}
                }

                log("[System] 发现 PID: $pid，等待 2.5 秒初始化...")
                Thread.sleep(2500) 
                
                executeInjectCommand(packageName, pid, tempAgentPath)
            }

            launchIntent?.let {
                it.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(it)
            }
        } else {
            val pid = getPid(packageName)
            if (pid != null) {
                executeInjectCommand(packageName, pid, tempAgentPath)
            }
        }
    }

    private fun executeInjectCommand(packageName: String, pid: String, agentPath: String) {
        log("[System] 执行 17.9.1 最终版注入指令...")
        Shell.cmd("setenforce 0").exec()
        val command = "cat /dev/null | ${injector.path} --pid $pid --script $agentPath --runtime=qjs"
        log("[System] 指令: $command")
        val result = Shell.cmd(command).exec()
        if (result.isSuccess) {
            log("[Success] 注入成功！脚本已加载 (QuickJS)")
        } else {
            log("[Error] 注入失败，错误码: ${result.code}")
            result.err.forEach { log("[frida-err] $it") }
            result.out.forEach { log("[frida-out] $it") }
            if (result.code == 4) {
                log("[Critical] 错误码 4 重现！请检查：")
                log("1. SELinux 状态: 执行 'getenforce' 应为 Permissive")
                log("2. 文件权限: /data/local/tmp/wrapped_agent.js 是否可读?")
                log("3. 尝试手动执行上述命令查看完整错误")
            }
        }
    }

    private fun log(msg: String) {
        loggingCallback?.invoke(msg)
    }

    class Builder(val context: Context) {
        private var armBinaryPath: String? = null
        private var arm64BinaryPath: String? = null
        private var x86BinaryPath: String? = null
        private var x86_64BinaryPath: String? = null
        var injector: File? = null
            private set

        fun withArmInjector(assetName: String) = apply { armBinaryPath = assetName }
        fun withArm64Injector(assetName: String) = apply { arm64BinaryPath = assetName }
        fun withX86Injector(assetName: String) = apply { x86BinaryPath = assetName }
        fun withX86_64Injector(assetName: String) = apply { x86_64BinaryPath = assetName }

        @Throws(IOException::class)
        fun build(): FridaInjector {
            val selectedName = when (arch) {
                "arm" -> armBinaryPath
                "arm64" -> arm64BinaryPath
                "x86" -> x86BinaryPath
                "x86_64" -> x86_64BinaryPath
                else -> null
            } ?: throw RuntimeException("injector binary not provided for arch: $arch")

            injector = extractInjectorIfNeeded(context, selectedName)
            return FridaInjector(this)
        }
    }

    // ========== 新增：使用 Termux PTY 的注入方法 ==========
    fun injectWithTerminalSession(
        fridaAgent: FridaAgent,
        packageName: String,
        spawn: Boolean = false,
        onOutput: (String) -> Unit,
        onError: (String) -> Unit,
        onComplete: () -> Unit
    ) {
        val agentScript = StringBuilder(fridaAgent.wrappedAgent)
        // TODO: interfaces 处理 (与原有逻辑相同)
        
        val fridaAgentFile = File(fridaAgent.getFilesDir(), "wrapped_agent.js")
        Utils.writeToFile(fridaAgentFile, agentScript.toString())
        Shell.cmd("chmod 777 ${fridaAgentFile.path}").exec()
        
        if (spawn) {
            val launchIntent = context.packageManager.getLaunchIntentForPackage(packageName)
            Shell.cmd("am force-stop $packageName").exec()
            launchIntent?.let {
                it.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(it)
            }
            thread {
                var pid: String? = null
                while (pid == null) {
                    Thread.sleep(500)
                    pid = getPid(packageName)
                }
                Thread.sleep(1500)
                executeInjectWithTerminalSession(packageName, pid, fridaAgentFile.path, onOutput, onError, onComplete)
            }
        } else {
            val pid = getPid(packageName)
            if (pid != null) {
                executeInjectWithTerminalSession(packageName, pid, fridaAgentFile.path, onOutput, onError, onComplete)
            } else {
                onError("[Error] 进程未运行: $packageName")
                onComplete()
            }
        }
    }

// 修改 executeInjectWithTerminalSession 方法
private fun executeInjectWithTerminalSession(
    packageName: String,
    pid: String,
    agentPath: String,
    onOutput: (String) -> Unit,
    onError: (String) -> Unit,
    onComplete: () -> Unit
) {
    val command = "${injector.path} -p $pid -s $agentPath -R qjs"
    onOutput("[System] 使用 PTY 终端执行: $command")
    
    val terminalSession = FridaTerminalSession(
        context = context,
        command = command,
        args = emptyArray(),
        onOutput = { line ->
            val cleanLine = line.replace(Regex("\\u001B\\[[;\\d]*[A-Za-z]"), "")
            if (cleanLine.isNotBlank()) onOutput(cleanLine)
        },
        onError = onError,
        onExit = { exitCode ->
            if (exitCode == 0) {
                onOutput("[Success] 注入完成")
            } else {
                onError("[Error] 注入失败，退出码: $exitCode")
            }
            onComplete()
        }
    )
    
    // 必须在主线程创建 TerminalSession（因为它内部创建了 Handler）
    Handler(Looper.getMainLooper()).post {
        terminalSession.start()
    }
}
}